### DISCUSSION TITLE
Evaluate using Profile-Guided Optimization (PGO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [tock/tock#3850] [Ideas]
Hi!

Recently I found that Profile-Guided Optimization (PGO) helps in many software domains to optimize CPU workloads - all my findings can be found [here](https://github.com/zamazan4ik/awesome-pgo).

One of many examples where PGO is already tested is the operating system domains. There are multiple examples:

* [Linux kernel](https://kernel.org/):
  - [Ugur Paper](https://web.eecs.umich.edu/~takh/papers/ugur-one-profile-fits-all-osr-2022.pdf)
  - [Microsoft presentation](https://lpc.events/event/7/contributions/771/attachments/630/1193/Exploring_Profile_Guided_Optimization_of_the_Linux_Kernel.pdf)
  - [ASOS (Application Specific Operating System)](http://scis.scichina.com/en/2018/092102.pdf)
  - [TCP Stream perf](https://lpc.events/event/7/contributions/798/attachments/661/1214/LTO_PGO_and_AutoFDO_-_Plumbers_2020_-_Tolvanen_Wendling_Desaulniers.pdf)
  - [Phoronix post](https://www.phoronix.com/news/Clang-PGO-For-Linux-Next)
  - Yet another attempt to PGO Linux kernel: http://coolypf.com/kpgo.htm
  - [Gentoo Wiki](https://wiki.gentoo.org/wiki/Kernel/Optimization#Performance)
  - Optimizing Linux kernel with Clang. An [article](https://habr.com/ru/companies/ruvds/articles/696236/)(in Russian) and [results](https://github.com/h0tc0d3/linux_pgo)
* Windows: 5-20% improvement according to the [presentation](https://lpc.events/event/7/contributions/771/attachments/630/1193/Exploring_Profile_Guided_Optimization_of_the_Linux_Kernel.pdf)

The Linux kernel already has ready-to-use KBUILD flags for building the kernel with PGO.

I think evaluating PGO for Tock can be a valuable thing too since PGO can help in the following ways:

* Better optimize CPU-intensive routines. It can lower requirements for CPU and reduce battery usage
* Can help with reducing the binary size (since the compiler can less frequently inline "cold" functions according to a PGO profile)

I can suggest trying the following things:

* Build Tock with PGO and perform all required benchmarks.
* If PGO shows some improvement - add a note to the Tock's documentation about building with PGO, and add corresponding build options for building Tock with enabled PGO (some inspirations from the Linux kernel or other PGO-enabled projects can be taken)

**Comment by bradjc** [comment]
Seems interesting!

Would it be difficult to generate the profiles on an embedded device?

**Reply by zamazan4ik** [reply]
> Would it be difficult to generate the profiles on an embedded device?

It depends on your device. If you have a filesystem in your device - it shouldn't be difficult. The PGO profile is usually a simple file on a disk. You can compile OS with instrumentation, run it, and then collect PGO profiles from a filesystem.

If you don't have an FS - it's still possible to do. Compilers have special intrinsics for dumping PGO profiles into a memory (e.g. for Clang you can check them [here](https://github.com/llvm/llvm-project/blob/main/compiler-rt/lib/profile/InstrProfiling.h)).

**Comment by jrvanwhy** [comment]
We'd be happy to accept contributions that help minimize the size and memory usage of our generated binaries!

If you're interested in helping out, we have some tools to measure flash and SRAM usage, such as [`print_tock_memory_usage.py`](https://github.com/tock/tock/blob/master/tools/print_tock_memory_usage.py).
